### DISCUSSION TITLE
CRT mem setup for level loading

### DISCUSSION THREAD

**Discussion by adaumann** [drmortalwombat/oscar64#60] [General]
Hello,
maybe someone could help me with my memory setup for a level loading from cartridge, my idea is to have a easyflash or magicdesk setup that all levels are written in C code and are on different banks. One bank can have n levels (i.e. 0x0800 bytes length max). The levels could have a fixed position on the bank (i.e. 0x8000, x8800, x9000 etc.) or the levels have a variable size (which is even better). I would like to have a largemem setup 0xa00-0xd000 in general. I would like to copy the level on a fixed position 0x7000 and execute it. I could not figure out how to make the correct setup:

I tried this, but this is not correct, but maybe gives you an idea:

#pragma section( bcode1, 0 )
#pragma section( bdata1, 0 )
#pragma section( bcode2, 0 )
#pragma section( bdata2, 0 )
#pragma region(bank1, 0x8000, 0xa000, , 1, { bcode1, bdata1, bcode2, bdata2 }, 0x2000 )

#pragma code ( bcode1 )
#pragma data ( bdata1 )

// Print into shared charwin

void print1(void)
{
	cwin_put_string(&cw, p"This is first bank 1", 1);
	cwin_cursor_newline(&cw);
}

const EntryTable	entry1 = {
	.fhello = &print1,
};

#pragma reference(entry1)

#pragma code ( bcode2 )
#pragma data ( bdata2 )

void print2(void)
{
	cwin_put_string(&cw, p"This is first bank 2", 1);
	cwin_cursor_newline(&cw);
}

const EntryTable	entry2 = {
	.fhello = &print2,
};

#pragma reference(entry2)

#pragma code ( code )
#pragma data ( data )

void callLevel(char bank, char level)
{
	// How to get the code address?
	memcpy((char *)0x7000, (char *)0x8000, 0x1000);

	// call function at start of copied section
	((EntryTable *)0x7000)->fhello();
}

**Comment by drmortalwombat** [comment]
You could use inlays to expand the level sections: https://github.com/drmortalwombat/oscar64?tab=readme-ov-file#inlays

**Comment by drmortalwombat** [comment]
I just added a sample for the inlays: https://github.com/drmortalwombat/OscarTutorials/tree/main/4530%20InlayLevels

**Reply by adaumann** [reply]
Wow, that was quick… thank you. I will try it out later. I need to figure it out how it works with Easyflash. My memory is running out… instead I could disk loading but CRT are a better option these days…

**Comment by adaumann** [comment]
I tried to set the inlay data on the easyflash, but the program crashes: What did I wrong?

``
#include <oscar.h>
#include <stdio.h>
#include <c64/memmap.h>
#include <c64/cia.h>
#include <c64/easyflash.h>

// The constant arrays that hold the compressed inlay regions
#pragma section( bdata1, 0 )
#pragma region(bank1, 0x8000, 0xc000, , 1, { bdata1 } )

#pragma data ( bdata1 )

const char Inlay0[];
const char Inlay1[];
const char Inlay2[];
const char Inlay3[];

#pragma data ( data )


// The inlay sections
#pragma section( icode0, 0 )
#pragma section( icode1, 0 )
#pragma section( icode2, 0 )
#pragma section( icode3, 0 )

// The regions to be compressed into the inlay constants
#pragma region( isec0, 0xc000, 0xd000, , Inlay0, { icode0 } )
#pragma region( isec1, 0xc000, 0xd000, , Inlay1, { icode1 } )
#pragma region( isec2, 0xc000, 0xd000, , Inlay2, { icode2 } )
#pragma region( isec3, 0xc000, 0xd000, , Inlay3, { icode3 } )


// Structure specifying a level.
struct Level
{
	const char * 	inlay;	// Inlay to expand for the level
	void (* entry)(void);	// Function to call in the inlay
};

// Switch to section for first inlay
#pragma code( icode0 )

// Entry function of first inlay
void Entry0(void)
{
	puts(p"entry inlay zero\n");
}

// Switch to section for second inlay
#pragma code( icode1 )

void Entry1(void)
{
	puts(p"entry inlay one\n");
}

#pragma code( icode2 )

void Entry2(void)
{
	puts(p"entry inlay two\n");
}

#pragma code( icode3 )

void Entry3(void)
{
	puts(p"entry inlay three\n");
}

// Now back to shared code
#pragma code( code )

// Constand structure with the level information
const Level	levels[] = {
	{Inlay0, Entry0},
	{Inlay1, Entry1},
	{Inlay2, Entry2},
	{Inlay3, Entry3}
};

int main(void)
{
	// Enable ROM
	mmap_set(MMAP_ROM);

	// Init CIAs (no kernal rom was executed so far)
	cia_init();

	// Loop over all levels
	for(char i=0; i<4; i++)
	{
        eflash.bank = 1;       
		// Expand the inlay of the level
		oscar_expand_lzo((char *)0xc000, levels[i].inlay);

		// Call the entry function of the level
		levels[i].entry();
	}
	return 0;
}
``

**Comment by drmortalwombat** [comment]
"puts()" uses the kernal for io, so this will be an issue.

**Comment by drmortalwombat** [comment]
To get the inlays into the bank 1 of the easyflash is a little tricky.  An unitialized variable goes per default to the bss section.  Specifying it later as an inlay will move it to the code section.  So correct would be:
`// The inlay sections
#pragma section( icode0, 0 )
#pragma section( icode1, 0 )
#pragma section( icode2, 0 )
#pragma section( icode3, 0 )

#pragma code ( bdata1 )

// The regions to be compressed into the inlay constants
#pragma region( isec0, 0xc000, 0xd000, , Inlay0, { icode0 } )
#pragma region( isec1, 0xc000, 0xd000, , Inlay1, { icode1 } )
#pragma region( isec2, 0xc000, 0xd000, , Inlay2, { icode2 } )
#pragma region( isec3, 0xc000, 0xd000, , Inlay3, { icode3 } )

#pragma code ( code )`

**Reply by adaumann** [reply]
I think I will follow a different route for using easyflash, that will give me the full memory (up to 1MB). I used your example easyflashshared.c and replaced the callback proxy function with my own version which has a stack for the latest bank. The result is a automatic bank switching and I could call every function from any bank and the bank is restored while return:

(a) Own Stack

#define SIZE 16


```
sbyte top;
byte inp_array[SIZE];
void push(byte x);
void pop();
byte popx;

void push(byte x)
{
    if (top == SIZE - 1)
	{
		return;
	}
    {
        top = top + 1;
        inp_array[top] = x;
    }
}

void pop()
{
	byte x;

    if (top == -1)
    {
        return;
    }
    else
    {
        popx = inp_array[top];
        top = top - 1;
   }
}

```

(b) callstack

```
__export void callbank(void (* fn)(void), char fromBank, char toBank)
{
	push(fromBank);
	eflash.bank = toBank;
	fn();
	pop();
	if(popx == 0xff)
	{
		vic.color_border = VCOL_LT_BLUE;
	}
	if(popx != 0xff)
	{
		eflash.bank = popx;
	}
}

#define FCALL(c,f) callbank(f, __bankof(c), __bankof(f))


```

(c) Sample call

```
#pragma code ( bcode2 )
#pragma data ( bdata2 )

void print2(void)
{
	byte y;
	y++;
	cwin_put_char(&cw, y, 7);

	cwin_put_string(&cw, p"This is second bank", 7);
	cwin_cursor_newline(&cw);
	FCALL(print2, print1);
	FCALL(print2, print3);
	// callbank(print1, 2, 1);
	// callbank(print3, 2, 3);
}


#pragma code ( bcode4 )
#pragma data ( bdata4 )

```

**Comment by drmortalwombat** [comment]
You can use __bankof("") to get the bank of the current function.  I am using a similar (but template based) method in https://github.com/drmortalwombat/Ninia to get from bank to bank.

**Comment by adaumann** [comment]
Hello, unfortunatly I could not get it to run. I sorted the functions into 3 different banks with minimal cross-calls  (game logic, actors, scenes) and the variables and common functions (also from imported libraries) are in a000-8000. The bank execution is with ROM on, the display functions are NO_ROM. As soon as I run code in the bank it crashes inbetween, I think while accessing global variables. I have no clue what is happing and debugging of crt is also not possible. 
I can try to put everything in main memory (with loading scenes from disk) but the memory will not be sufficient I assume. 
If you are interested to solve the memory management please let me know and I can provide you the sources or I need to find a different solution if possible.

**Reply by drmortalwombat** [reply]
I would like to see this in action, so please share your code.  You can also find me on the Everything C64 discord, if you want to chat.

**Comment by adaumann** [comment]
Here is the source code: This contains a working file in main memory only and a easyflash try. 


It would be nice if you could tell what the issue is with easyflash. Discord I save for later...

**Reply by drmortalwombat** [reply]
Main problem with crashing is the interrupt.  If you enable the ROM you have to go through the kernal IRQ multiplexer `rirq_init(true);`  and you also have to enable the trampoline for the points in time where there is no ROM banked in.

**Reply by adaumann** [reply]
Ok, after I disable raster interrupts it works again, but the bitmap functions have strange artifacts. I ignore that first. Now: Because Easyflash needs the ROM enabled and some drawings need no kernal enabled I need to switch it constantly. How can I change the rasterinterrupt between two memory settings? Or should I move the bitmap video location in 0x2000, but than the memory from e000-ffff is not usable... Do you have some ideas

Thanks for the fast reply...

**Reply by drmortalwombat** [reply]
If you use the trampoline the interrupt will work with ROM or no ROM when using the kernal IRQ multiplexer.  The trampoline enables the ROM, executes the interrupt and restores the previous setting.

You can copy one ROM bank from 8000 to bfff to RAM and use it with ROM disabled.  If you put all raster operations into that bank you should be fine.

**Reply by adaumann** [reply]
Now it is working with easyflash. Thank you for the hint with the "trampoline", I was not aware about the real function. There is one problem after I switch to: "rirq_init(true)" with the trampoline, I think this is bug in the bitmap library. The graphics have some strange artifacts. See attachment  
![image](https://github.com/drmortalwombat/oscar64/assets/94932945/49358cc8-7a03-41c5-b1ea-aca7db919abf)

**Reply by drmortalwombat** [reply]
Probably conflicting with the kernal ROM, if the bitmap is $e000..$ffff.  You can write through the ROM but not read the RAM with the ROM enabled.  So anything that does read-modify-write or masking will generate artifacts.

**Comment by adaumann** [comment]
Sounds logically, I need to check my code, probably I read bitmap patterns while Rom is enabled… so I’m glad that the easyflash version is running… gives me up to 1MB 🥳 with bank switching. thanks

**Reply by adaumann** [reply]
I solved the bitmap issues by using these kind of proxy functions in the main ram: 

// Call a bank 0 function from any other bank
void proxy0_deleteRender(byte id)
{
	mmap_trampoline();
	mmap_set(MMAP_NO_ROM);
	deleteRender(id);
	mmap_set(MMAP_ROM);
}

// Call a bank2 function from bank 1
void proxy12_dispatchEvent(byte event, byte senderId, byte p1, byte p2)
{
	eflash.bank = 2;
	dispatchEvent_2(event, senderId, p1, p2);
	eflash.bank = 1;
}

BTW: Thank you for providing the C compiler and the rich library, great work for an old Commodore !

**Reply by drmortalwombat** [reply]
You should only install the trampoline once at program start

**Reply by adaumann** [reply]
ok, but I need a rirq_wait(); before switching to no_rom... otherwise it might crash.
